### PR TITLE
Enable mkFit in InitialStepPreSplitting,InitialStep,HighPtTripletStep,DetachedQuadStep tracking iterations for phase1 pixel era, except for HI and special 2017 tracking eras

### PR DESCRIPTION
**Pull request by makortel**
#### PR description:

This PR enables mkFit by default in the following 4 tracking iterations
* InitialStepPreSplitting
* InitialStep
* HighPtTripletStep
* DetachedQuadStep

for the entire era of phase 1 pixel detector (i.e. from 2017 to end of Run 3), except for HI eras (`Run2_2017_pp_on_XeXe`, `Run2_2017_ppRef`, `Run2_2018_pp_on_AA`, `Run2_2018_pp_on_AA_noHCALmitigation`, `Run3_pp_on_PbPb`) and 2017 special tracking eras (`Run2_2017_trackingLowPU`, `Run2_2017_trackingRun2`), following the plan presented e.g. in https://indico.cern.ch/event/1071296/#1-news .

This PR also adds `Run3_noMkFit` Era to easily use Run3 era without mkFit.

Technically this PR makes use of the existing per-iteration Modifiers in order to avoid conflicts with concurrent developments to enable mkFit in additional iterations. Those are planned to be cleaned up soon, after the development settles.

#### PR validation:

Checked locally that both 11634.0 and 11634.7 run, and from their printouts that mkFit is used in initialStep.

Expecting no changes in 11634.7. Expecing changes in all (other) phase 1 pixel workflows (except those eras listed above) that are along those shown in https://indico.cern.ch/event/992975/?showDate=all&showSession=all#8-mkfit-status-and-plans.

### CODE REVIEW

**Comment by makortel** [comment]
test parameters:
   - workflows = 148.0,149.0,140.56,159.0,10024.3,10024.4

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by makortel** [comment]
assign reconstruction

File: Configuration/Eras/python/Era_Run2_2017_cff.py [hunk 1]
```diff
@@ -26,5 +27,5 @@
 
 Run2_2017 = cms.ModifierChain(Run2_2016.copyAndExclude([run2_muon_2016, run2_HLTconditions_2016, run2_ECAL_2016, run2_egamma_2016,pixel_2016,run2_jme_2016, strips_vfp30_2016, ctpps_2016]),
                               phase1Pixel, run2_ECAL_2017, run2_HF_2017, run2_HCAL_2017, run2_HE_2017, run2_HEPlan1_2017, 
-                              trackingPhase1, run2_GEM_2017, stage2L1Trigger_2017, run2_HLTconditions_2017, run2_muon_2017,run2_egamma_2017, ctpps_2017, run2_jme_2017)
+                              trackingPhase1, trackingMkFitProd, run2_GEM_2017, stage2L1Trigger_2017, run2_HLTconditions_2017, run2_muon_2017,run2_egamma_2017, ctpps_2017, run2_jme_2017)
```
[the remark this answers is not in the document]
Comment (makortel) [reply Configuration/Eras/python/Era_Run2_2017_cff.py:30]: Thanks, that could indeed make sense.

**Comment by makortel** [comment]
> ---> test TestConfigDP had ERRORS

@cms-sw/reconstruction-l2 @cms-sw/operations-l2 The failure comes from
```
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_1_X_2021-09-28-1100/src/Configuration/DataProcessing/test/RunExpressProcessing.py --scenario HeavyIonsEra_Run2_2018 --global-tag GLOBALTAG --lfn /store/whatever --fevt --dqmio --alcareco TkAlMinBiasHI+SiStripCalMinBias
Retrieved Scenario: HeavyIonsEra_Run2_2018
...
Traceback (most recent call last):
  File "/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_1_X_2021-09-28-1100/src/Configuration/DataProcessing/test/RunExpressProcessing.py", line 97, in __call__
    process = scenario.expressProcessing(self.globalTag, **kwds)
  File "/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_1_X_2021-09-28-1100/python/Configuration/DataProcessing/Impl/HeavyIons.py", line 88, in expressProcessing
    process = Reco.expressProcessing(self,globalTag, **args)
...
  File "/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_1_X_2021-09-28-1100/python/Configuration/StandardSequences/ReconstructionHeavyIons_cff.py", line 30, in <module>
...
```
that point to
https://github.com/cms-sw/cmssw/blob/1b0b930c3fcf8dfdc8334d2e54b62fdfd3e408af/Configuration/DataProcessing/python/Impl/HeavyIonsEra_Run2_2018.py#L18-L21
https://github.com/cms-sw/cmssw/blob/1b0b930c3fcf8dfdc8334d2e54b62fdfd3e408af/Configuration/DataProcessing/python/Impl/HeavyIons.py#L16-L20
https://github.com/cms-sw/cmssw/blob/1b0b930c3fcf8dfdc8334d2e54b62fdfd3e408af/Configuration/Applications/python/ConfigBuilder.py#L1056-L1063

So the 2018 HI run with the `Run2_2018` + `ReconstructionHeavyIons_cff`? I'm not really sure what to do here. Easiest would be to not enable mkFit for 2018 (and maybe for 2017 too) for now.

**Comment by slava77** [comment]
> So the 2018 HI run with the `Run2_2018` + `ReconstructionHeavyIons_cff`? I'm not really sure what to do here. Easiest would be to not enable mkFit for 2018 (and maybe for 2017 too) for now.

it should be OK to remove `HeavyIonsEra_Run2_2018` (HI switched to pp-based scenario  in 2018 and the HeavyIons variant was left just in case), but I think that there is still something similar in runTheMatrix exercised in the IBs.

It would make more sense to add a heavyIons modifier to have more freedom overall.

File: Configuration/Eras/python/Era_Run2_2017_cff.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (makortel) [reply Configuration/Eras/python/Era_Run2_2017_cff.py:30]: On closer inspection, the problem is that
* 2017 "not standard pp" Eras derive from `Run2_2017`, so they need the exclusions
* 2018 "not standard pp" Eras derive from `Run2_2018` (that itself derives from `Run2_2017`), so they need the exclusions
* Run 3 "not standard pp" Eras derive from `Run3` (that derives from `Run2_2018`), so they need the exclusions
* `Phase2` derives from `Run3`, so it needs the exclusion

A (tree) structure of the Eras where intermediate ModifierChains would be used e.g. for
* 2017 modifiers that should carry over to 2018
* 2017 modifiers that should not carry over but are common for all 2017 eras
* 2018 modifiers that should carry over to Run 3
* 2018 modifiers that should not carry over but are common for all 2018 eras
* Run 3 (2022?) modifiers that should carry over to Phase 2 (or really 2023?)
* Run 3 (2022?) modifiers that should not carry over but are common for all Run 3 (or really 2023?) eras

would be cleaner as the year-specific modifiers would not have to be excluded from the Era of the following year, but I think such reorganization is out of scope of this PR.

(on a historical note, such tree-like structure was the original intention to craft ModifierChains, until me, among others, asked loudly for `copyAndExclude()`)

**Comment by makortel** [comment]
> > So the 2018 HI run with the `Run2_2018` + `ReconstructionHeavyIons_cff`? I'm not really sure what to do here. Easiest would be to not enable mkFit for 2018 (and maybe for 2017 too) for now.
> 
> it should be OK to remove `HeavyIonsEra_Run2_2018` (HI switched to pp-based scenario in 2018 and the HeavyIons variant was left just in case), but I think that there is still something similar in runTheMatrix exercised in the IBs.
> 
> It would make more sense to add a heavyIons modifier to have more freedom overall.

I ended up making a "guaranteed CKF" module available for affected iterations that the HI configurations then clone. (I have vague recollections doing similar thing for the CA at the time for 2017)

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by slava77** [comment]
based on the discussion in the PPD today I understood that there will be a need to make incremental comparisons with CKF at submission with a modifier; that's assuming that this PR is merged to pre4. So, perhaps a `Run3_noMkFit` era would be useful for that test.

@rappoccio please check/confirm.

**Comment by rappoccio** [comment]
@slava77 Yes, there was a general consensus that a modifier would be helpful here. I think we should add one.

**Comment by makortel** [comment]
Ok, I'll add `Run3_noMkFit` Era in the next update (that will fix that and some further HI workflow failures).

**Comment by makortel** [comment]
test parameters:
   - workflows = 140.0,140.51,140.52,140.53,140.54,140.55,140.56,140.5611,140.57,145.0,148.0,149.0,150.0,158.0,159.0,10024.3,10024.4

**Comment by makortel** [comment]
(I hope this is a representative-enough set of HI workflows)

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by jpata** [comment]
I have no comments on the code. I want to check the outcome of the discussion today at
- BTV: https://indico.cern.ch/event/1083393/#1-mkfit-validation
- JME: https://indico.cern.ch/event/1082284/#20-recent-validations-mkfitdee
- TRK: https://indico.cern.ch/event/1079748/#3-mkfit-status-report

before signing.

**Comment by jpata** [comment]
@emilbols any issue with going forward with this PR following the BTV meeting today?

**Comment by emilbols** [comment]
@jpata The btv conveners would like to do some additional checks, before approving since a small difference in tagging performance is observed, see [1].

[1]
https://cms-btv-pog-validation.web.cern.ch/cms-btv-pog-validation/validation/packages/RecoB/CMSSW_12_1_0_pre2/TTbar_1210pre2_tkmkfit_2021_14TeV_vs_1210pre2_2021_14TeV_FullSim_PU/deepCSV_BvsAll_performance_vs_B_GLOBAL_all.png

**Comment by jpata** [comment]
> The btv conveners would like to do some additional checks, before approving since a small difference in tagging performance is observed

Could you please clarify if this implies that the sign-off in [12_1_0_pre2_2021_TkmkFitRecoOnly](https://cms-pdmv.cern.ch/valdb/?srch=12_1_0_pre2&selected=12_1_0_pre2_2021_TkmkFitRecoOnly&Reconstruction=true&RData=true&RFull=true&RFast=true&HLT=true&HData=true&HFull=true&IN=true&IData=true&IFull=true&GEN=true&GGen=true&PAGs=true&PData=true&PFull=true&PFast=true) was redacted?

cc @cms-sw/pdmv-l2

**Comment by IHateLinus** [comment]
> > The btv conveners would like to do some additional checks, before approving since a small difference in tagging performance is observed
> 
> Could you please clarify if this implies that the sign-off in [12_1_0_pre2_2021_TkmkFitRecoOnly](https://cms-pdmv.cern.ch/valdb/?srch=12_1_0_pre2&selected=12_1_0_pre2_2021_TkmkFitRecoOnly&Reconstruction=true&RData=true&RFull=true&RFast=true&HLT=true&HData=true&HFull=true&IN=true&IData=true&IFull=true&GEN=true&GGen=true&PAGs=true&PData=true&PFull=true&PFast=true) was redacted?
> 
> cc @cms-sw/pdmv-l2

There has been quite a few improvements on (the already rather good pre2) and no real show-stoppers found so far. There are legitimate requests for larger statistics tests and those should be made. In discussions last week w Andrea and Sal, it was agreed that the next phase of validations and larger samples will be done AFTER and USING pre4.

**Comment by emilbols** [comment]
> Could you please clarify if this implies that the sign-off in [12_1_0_pre2_2021_TkmkFitRecoOnly](https://cms-pdmv.cern.ch/valdb/?srch=12_1_0_pre2&selected=12_1_0_pre2_2021_TkmkFitRecoOnly&Reconstruction=true&RData=true&RFull=true&RFast=true&HLT=true&HData=true&HFull=true&IN=true&IData=true&IFull=true&GEN=true&GGen=true&PAGs=true&PData=true&PFull=true&PFast=true) was redacted?

Yes, it was redacted

**Comment by jpata** [comment]
Thanks for the clarifications. So this looks like a PPD / PdmV question.

**Comment by jpata** [comment]
+reconstruction
- mkFit is enabled in certain tracking steps with a modifier, as discussed in the presentation linked in the description
- since the default is changed, there are expected reco changes in many quantities downstream from tracks
- this sig is for the code changes from the technical/reco point of view. The decision if and in which release to merge it is not for reco, and perhaps can be brought up at the ORP meeting.

**Comment by kskovpen** [comment]
Looks good to PdmV

**Comment by kskovpen** [comment]
+1

**Comment by perrotta** [comment]
+1
- Mkfit is enabled in pre4 for the first four tracking iterations, as decided right now at the ORP: a protection is put in place for HI workflows
- A modifier is defined here to allow switching back to the `Run3_noMkFit` case

**Comment by perrotta** [comment]
+operations

**Pull request merged**